neqo-transport/src/ecn.rs

```diff
@@ -0,0 +1,116 @@
+                self.state = EcnValidationState::Unknown;
+            }
+        }
+    }
+
+    /// After the ECN validation test has ended, check if the path is ECN capable.
+    pub fn validate_ack_ecn(
+        &mut self,
+        space: PacketNumberSpace,
+        acked_packets: &[SentPacket],
+        ecn_count: &EcnCount,
+    ) {
+        // RFC 9000, Appendix A.4:
+        // From the "unknown" state, successful validation of the ECN counts in an ACK frame
+        // (see Section 13.4.2.1) causes the ECN state for the path to become "capable", unless
+        // no marked packet has been acknowledged.
+        if self.state != EcnValidationState::Unknown {
+            return;
+        }
+
+        // RFC 9000, Section 13.4.2.1:
+        //
+        // > An endpoint that receives an ACK frame with ECN counts therefore validates
+        // > the counts before using them. It performs this validation by comparing newly
+        // > received counts against those from the last successfully processed ACK frame.
+        //
+        // RFC 9000 fails to state that this is done *per packet number space*.
+        //
+        // > If an ACK frame newly acknowledges a packet that the endpoint sent with
+        // > either the ECT(0) or ECT(1) codepoint set, ECN validation fails if the
+        // > corresponding ECN counts are not present in the ACK frame.
+        //
+        // We always mark with ECT(0) - if at all - so we only need to check for that.
+        // Also, if we sent a packet with ECT(0) and get only an ACK frame (and not an
+        // ACK-ECN frame), `ecn_counts` will be all zero and the check below will fail,
+        // so no need to explicitly check for the above.
+        //
+        // > ECN validation also fails if the sum of the increase in ECT(0) and ECN-CE counts is
+        // > less than the number of newly acknowledged packets that were originally sent with an
+        // > ECT(0) marking.
+        let newly_acked = acked_packets.len().try_into().unwrap();
+        let ecn_diff = ecn_count - &self.count[space];
+        let sum_inc = ecn_diff[IpTosEcn::Ect0] + ecn_diff[IpTosEcn::Ce];
+        if sum_inc < newly_acked {
+            qwarn!(
+                "ACK had {} new marks, but acked {} packets, ECN validation failed",
+                sum_inc,
+                newly_acked
+            );
+            self.state = EcnValidationState::Failed;
+        } else {
+            qinfo!("ECN validation succeeded, path is capable");
+            self.state = EcnValidationState::Capable;
+        }
+        self.count[space] = ecn_count.clone();
+    }
+
+    /// The ECN mark to use for packets sent on this path.
+    pub fn ecn_mark(&self) -> IpTosEcn {
+        match self.state {
+            EcnValidationState::Testing | EcnValidationState::Capable => IpTosEcn::Ect0,
+            EcnValidationState::Failed | EcnValidationState::Unknown => IpTosEcn::NotEct,
+        }
+    }
```

Comment:
When n `EcnValidationState::Unknown`, one should still mark with `Ect0`, no?

Otherwise `sum_inc < newly_acked` check above triggers, where a returning ACK acknowledges more packets than it has marks.

```suggestion
    /// The ECN mark to use for packets sent on this path.
    pub fn ecn_mark(&self) -> IpTosEcn {
        match self.state {
            EcnValidationState::Testing | EcnValidationState::Capable | EcnValidationState::Unknown => IpTosEcn::Ect0,
            EcnValidationState::Failed => IpTosEcn::NotEct,
        }
    }
```

Reply:
That's not how I understand the text in RFC9000 (but that text is far from clear). I think the assumption is that one would receive an ACK within those ten packets that would validate or invalidate the ECN capability of the path.

Reply:
> When in `EcnValidationState::Unknown`, one should still mark with `Ect0`, no?

How would we recover from the case where a path drops all packets with ECT(0)?